chef/cookbooks/crowbar/templates/default/crowbar-stop-nova-services.sh.erb

```diff
@@ -0,0 +1,88 @@
+#!/bin/bash
+#
+# After the upgrade of all compute services on all nodes is finished, it's
+# necessary to stop all nova services so that they start using the latest RPC
+# API version. All services must be stopped without exception: if any service
+# still uses the old RPC API version, it may cause other services to pick that
+# RPC API version as well.
+
+LOGFILE=/var/log/crowbar/node-upgrade.log
+UPGRADEDIR=/var/lib/crowbar/upgrade
+SCRIPTNAME=stop-nova-services
+mkdir -p "`dirname "$LOGFILE"`"
+exec >>"$LOGFILE" 2>&1
+
+log()
+{
+    set +x
+    echo "[$(date --iso-8601=ns)] [$$] $@"
+    set -x
+}
+
+log "Executing $BASH_SOURCE"
+
+set -x
+
+if [[ -f $UPGRADEDIR/crowbar-${SCRIPTNAME}-ok ]] ; then
+    log "Stop nova script was already successfully executed"
+    exit 0
+fi
+
+<% if @nova_controller %>
+for service in conductor scheduler novncproxy serialproxy api; do
+    fullname="openstack-nova-$service"
+    if systemctl --quiet is-active $fullname 2>/dev/null ; then
+        systemctl stop $fullname
+    fi
+done
+
+# This is needed to prevent the Nova Placement API from causing RPC version
+# trouble.
+if systemctl --quiet is-active apache2 2>/dev/null ; then
+    systemctl stop apache2
+fi
+
+  <% if @remotes_present and @is_cluster_founder %>
+# Stop nova-compute on remote nodes:
+for remote in $(crm resource list | awk '$1 ~ /^remote-/  {print $1}'); do
+  crm resource stop $remote
+done
+  <% end %>
+
+# Remove and unmanage openstack-nova-consoleauth
+systemctl disable openstack-nova-consoleauth
+systemctl stop openstack-nova-consoleauth
+systemctl kill openstack-nova-consoleauth
+rpm -e openstack-nova-consoleauth
+
+<% else %>
+  # Only stop on non-remote compute nodes - service on remote nodes will be stopped by
+  # the controller side code above.
+  <% unless @is_remote_node %>
+systemctl stop openstack-nova-compute
+  <% end %>
+<% end %>
+
+# This check can run on both remote and non-remote nodes: the `crm resource
+# stop` for the remote resources will have been run before this script runs on
+# the compute node it was intended for. This check needs to fail on the compute
+# node itself if it is going to fail. That way it will pinpoint the problematic
+# node.
+
+for wait in 10 20 40 0; do
+  if [ $(ps -U nova u | wc -l) -gt 1 ]; then
```

Comment:
This is actually still wrong:

```
# ps -U nova u 
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
nova     28717  0.0  0.1  36912  4656 ?        Ss   14:24   0:00 /usr/lib/systemd/systemd --user
nova     28720  0.0  0.0 166136  1980 ?        S    14:24   0:00 (sd-pam)
```

Reply:
Sigh. This is a veritable Rubik's Cube :-)

Modified again...and that looks like it does the trick on both controller...

```
root@d00-25-90-e5-04-54:~ # ps -U nova u | grep -e apache -e python
nova      3811  0.0  0.5 307760 95552 ?        Ss   Oct30   0:17 /usr/bin/python /usr/bin/nova-novncproxy
nova      8240  0.0  0.6 593152 111972 ?       Sl   Oct29   0:28 (wsgi:nova-placement-ap -DSYSCONFIG -C PidFile /var/run/httpd.pid -C Include /etc/apache2/sysconfig.d//loadmodule.conf -C Include /etc/apache2/sysconfig.d//global.conf -f /etc/apache2/httpd.conf -c Include /etc/apache2/sysconfig.d//include.conf -DSYSTEMD -DFOREGROUND -k start
nova      8241  0.0  0.6 593144 110904 ?       Sl   Oct29   0:27 (wsgi:nova-placement-ap -DSYSCONFIG -C PidFile /var/run/httpd.pid -C Include /etc/apache2/sysconfig.d//loadmodule.conf -C Include /etc/apache2/sysconfig.d//global.conf -f /etc/apache2/httpd.conf -c Include /etc/apache2/sysconfig.d//include.conf -DSYSTEMD -DFOREGROUND -k start
nova      8242  0.0  0.6 658704 111116 ?       Sl   Oct29   0:28 (wsgi:nova-placement-ap -DSYSCONFIG -C PidFile /var/run/httpd.pid -C Include /etc/apache2/sysconfig.d//loadmodule.conf -C Include /etc/apache2/sysconfig.d//global.conf -f /etc/apache2/httpd.conf -c Include /etc/apache2/sysconfig.d//include.conf -DSYSTEMD -DFOREGROUND -k start
nova      8291  1.1  0.6 274660 100344 ?       Ss   Oct29  27:12 /usr/bin/python /usr/bin/nova-conductor
nova      8517  0.0  0.6 288568 113976 ?       Ss   Oct29   2:02 /usr/bin/python /usr/bin/nova-scheduler
nova      8624  1.1  0.7 311244 130692 ?       Ss   Oct29  27:26 /usr/bin/python /usr/bin/nova-api
nova      8807  0.0  0.6 284080 109700 ?       Ss   Oct29   1:41 /usr/bin/python /usr/bin/nova-consoleauth
nova     24523  0.0  0.8 323888 137768 ?       S    Oct30   0:00 /usr/bin/python /usr/bin/nova-api
nova     24527  0.0  0.8 322960 136960 ?       S    Oct30   0:00 /usr/bin/python /usr/bin/nova-api
nova     24528  0.0  0.8 323924 137784 ?       S    Oct30   0:01 /usr/bin/python /usr/bin/nova-api
nova     24529  0.0  0.8 324360 137868 ?       S    Oct30   0:00 /usr/bin/python /usr/bin/nova-api
nova     24530  0.0  0.8 318868 131488 ?       S    Oct30   0:00 /usr/bin/python /usr/bin/nova-api
nova     24531  0.0  0.8 318868 131488 ?       S    Oct30   0:00 /usr/bin/python /usr/bin/nova-api
nova     24532  0.0  0.8 318868 131488 ?       S    Oct30   0:00 /usr/bin/python /usr/bin/nova-api
nova     24533  0.0  0.8 318868 131488 ?       S    Oct30   0:00 /usr/bin/python /usr/bin/nova-api
nova     24577  0.2  0.7 297400 117364 ?       S    Oct30   2:07 /usr/bin/python /usr/bin/nova-conductor
nova     24578  0.2  0.7 297476 117592 ?       S    Oct30   2:08 /usr/bin/python /usr/bin/nova-conductor
nova     24585  0.2  0.7 296896 116796 ?       S    Oct30   2:05 /usr/bin/python /usr/bin/nova-conductor
nova     24586  0.2  0.7 297032 116952 ?       S    Oct30   2:09 /usr/bin/python /usr/bin/nova-conductor
root@d00-25-90-e5-04-54:~ #
```

...and compute node:

```
root@d00-25-90-e5-0c-fc:~ # ps -U nova u | grep -e apache -e python
nova     13415  0.2  0.7 1961284 130156 ?      Ssl  Oct30   2:30 /usr/bin/python /usr/bin/nova-compute
nova     13667  0.2  0.7 1961524 129976 ?      Ssl  Oct30   2:28 /usr/bin/python /usr/bin/nova-compute
root@d00-25-90-e5-0c-fc:~ #
```